File: rc2021-11-19.cpp
// rc2021-11-19.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>


// The Python version of the code looked something like this:
// for raw_strat in itertools.product(range(3), repeat=27):
//      strat_0 = chunks(raw_strat[0:9], 3)
//      strat_1 = chunks(raw_strat[9:18], 3)
//      strat_2 = chunks(raw_strat[18:27], 3)
// where chunks(lst, n) reshapes lst into a 3-column 2D nested list.
// Each strat is a 3x3 nested list where, e.g., strat_0[l][r] is what that strategy
// says you should guess when the color of the left hat is l and the color of the
// right hat is r.
//
// This version uses r[27] instead of raw_strat and uses this formula to determine
// which cell of the array to check:

inline int to_index(int strategy, int left, int right) {
    return (9 * strategy + 3 * left + right);
}

// This is where all the work happens.
// s0, s1, s2, and s3 are the numbers (from 0 to 2) of the strategy that player
// will use.
void iterate(int s0, int s1, int s2, int s3) {
	int r[27];
	int most_successes = -1;
	// From the symmetry of the problem, r[0] must be 0 and the first nonzero
	// digit must be 1. If I hadn't found an answer, I would've tried to skip
	// some of these cases.
	for (r[0] = 0; r[0] < 1; r[0]++) {
	for (r[1] = 0; r[1] < 2; r[1]++) {
	for (r[2] = 0; r[2] < 3; r[2]++) {
	for (r[3] = 0; r[3] < 3; r[3]++) {
	for (r[4] = 0; r[4] < 3; r[4]++) {
	for (r[5] = 0; r[5] < 3; r[5]++) {
	for (r[6] = 0; r[6] < 3; r[6]++) {
	for (r[7] = 0; r[7] < 3; r[7]++) {
	for (r[8] = 0; r[8] < 3; r[8]++) {
		// This prints every 12 or so seconds on my iMac in Windows 10.
		std::cout << "Checking ";
		for (int i = 0; i < 9; i++) std::cout << r[i];
		std::cout << "...\n";
		for (r[9] = 0; r[9] < 3; r[9]++) {
		for (r[10] = 0; r[10] < 3; r[10]++) {
		for (r[11] = 0; r[11] < 3; r[11]++) {
		for (r[12] = 0; r[12] < 3; r[12]++) {
		for (r[13] = 0; r[13] < 3; r[13]++) {
		for (r[14] = 0; r[14] < 3; r[14]++) {
		for (r[15] = 0; r[15] < 3; r[15]++) {
		for (r[16] = 0; r[16] < 3; r[16]++) {
		for (r[17] = 0; r[17] < 3; r[17]++) {
		for (r[18] = 0; r[18] < 3; r[18]++) {
		for (r[19] = 0; r[19] < 3; r[19]++) {
		for (r[20] = 0; r[20] < 3; r[20]++) {
		for (r[21] = 0; r[21] < 3; r[21]++) {
		for (r[22] = 0; r[22] < 3; r[22]++) {
		for (r[23] = 0; r[23] < 3; r[23]++) {
		for (r[24] = 0; r[24] < 3; r[24]++) {
		for (r[25] = 0; r[25] < 3; r[25]++) {
		for (r[26] = 0; r[26] < 3; r[26]++) {
			// Here's our array of hats. I internally matched RYB -> 012,
			// but that's completely arbitrary. Also, I put the players in
			// counterclockwise order, so their left is the same as the left
			// in a list written left to right.
			int h[4];
			for (h[0] = 0; h[0] < 3; h[0]++) {
			for (h[1] = 0; h[1] < 3; h[1]++) {
			for (h[2] = 0; h[2] < 3; h[2]++) {
			for (h[3] = 0; h[3] < 3; h[3]++) {
				// If nobody guessed right, get out of the hat loops and continue.
				if (h[0] != r[to_index(s0, h[3], h[1])] &&
						h[1] != r[to_index(s1, h[0], h[2])] &&
						h[2] != r[to_index(s2, h[1], h[3])] &&
						h[3] != r[to_index(s3, h[2], h[0])]) {
					goto SKIP; // My first goto in a C/C++ program in 30 years!
				} // end if
			}}}} // end h-loops
			// If we've made it here, we've checked every combination of hats
			// and someone's answered properly in each one.
			// Each section of code checks who ran each strategy,
			// then prints out that strategy.
			if (s0 == 0) std::cout << "0";
			if (s1 == 0) std::cout << "1";
			if (s2 == 0) std::cout << "2";
			if (s3 == 0) std::cout << "3";
			std::cout << ": ";
			for (int i = 0; i < 9; i++) std::cout << r[i];
			std::cout << "; ";
			if (s0 == 1) std::cout << "0";
			if (s1 == 1) std::cout << "1";
			if (s2 == 1) std::cout << "2";
			if (s3 == 1) std::cout << "3";
			std::cout << ": ";
			for (int i = 9; i < 18; i++) std::cout << r[i];
			std::cout << "; ";
			if (s0 == 2) std::cout << "0";
			if (s1 == 2) std::cout << "1";
			if (s2 == 2) std::cout << "2";
			if (s3 == 2) std::cout << "3";
			std::cout << ": ";
			for (int i = 18; i < 27; i++) std::cout << r[i];
			std::cout << " = 81\n";
			SKIP: ; // Is the null operation necessary? Beats me.
		} // end innermost r-loop
		}}}}}}}}}}}}}}}}} // end r-loops until we get to the progress indicator
	}}}}}}}}} // end rest of r-loops
} // end iterate()

int main()
{
	// Like I said, all the work is in the iterate() function. I thought it would
	// be better to have the repeated strategy change the most quickly, so I used
	// those arguments instead of iterate(0, 0, 1, 2).
	iterate(0, 1, 2, 2);
}

File: RC-2021-11-19.py
import itertools

#########
# This was lifted from Stack Overflow.

def chunks(lst, n):
    """Yield successive n-sized chunks from lst."""
    for i in range(0, len(lst), n):
        yield lst[i:i + n]

# Assuming "mathy" ccw order for the players:
#
#    1
#  2 @ 0
#    3

#########
# This was the first function I wrote. It gave everyone the same
# strategy. Its best result was shown above.
            
def test_all_same():
    most_successes = 0
    for raw_strat in itertools.product(range(3), repeat=9):
        strat = list(chunks(raw_strat, 3))

        successes = 0
        for hats in itertools.product(range(3), repeat=4):
            guesses = [strat[hats[3]][hats[1]],
                       strat[hats[0]][hats[2]],
                       strat[hats[1]][hats[3]],
                       strat[hats[2]][hats[0]]]
            if (hats[0] == guesses[0] or
                hats[1] == guesses[1] or
                hats[2] == guesses[2] or
                hats[3] == guesses[3]):
                    successes += 1

        if successes > most_successes:
            most_successes = successes
            print(strat, successes)

# Results:
# bs = best single strategy
#    = person to your left, if left != right
#      1 + person to left (mod 3), if left == right
bs = [[1,0,0], [1,2,1], [2,2,0]]

#########
# The best single strategy I found only failed if all the hats were the
# same color. I tried giving 3 players that strategy and testing every
# possible strategy for the 0th player. It found no successful strategies.

def test_3_best():
    """s0 iterates; s1~3 are all [[1,0,0],[1,2,1],[2,2,0]]"""
    most_successes = 0
    for raw_strat in itertools.product(range(3), repeat=9):
        strat = list(chunks(raw_strat, 3))

        successes = 0
        for hats in itertools.product(range(3), repeat=4):
            guesses = [strat[hats[3]][hats[1]],
                       bs[hats[0]][hats[2]],
                       bs[hats[1]][hats[3]],
                       bs[hats[2]][hats[0]]]
            if (hats[0] == guesses[0] or
                hats[1] == guesses[1] or
                hats[2] == guesses[2] or
                hats[3] == guesses[3]):
                    successes += 1
        if successes > most_successes:
            most_successes = successes
            print(strat, successes)
            
# Things were about to get a lot slower. I realized I'd need a hat test that
# just returned pass or fail, which would allow it to return fail as soon
# as possible. I made a more verbose version as well, which could
# help me logic through an answer. (Spoiler: it didn't.)

#########
# This function checks all 3**4 == 81 arrangements of hats, returning
# the number of successes. In verbose mode, it prints all the results
# and indicates which arrangements were only guessed correctly by
# one player.

def check_all_hats(s0, s1, s2, s3, verbose=False):
    """s0~3 = 3x3 strategy for that player"""
    successes = 0
    for hats in itertools.product(range(3), repeat=4):
        guesses = [s0[hats[3]][hats[1]],
                   s1[hats[0]][hats[2]],
                   s2[hats[1]][hats[3]],
                   s3[hats[2]][hats[0]]]
        if (hats[0] == guesses[0] or hats[1] == guesses[1] or
            hats[2] == guesses[2] or hats[3] == guesses[3]):
                successes += 1
                if verbose:
                    correct_bits = 0
                    for h in range(4):
                        if hats[h] == guesses[h]:
                            print(f'<{hats[h]}>', end='')
                            correct_bits += 2**h
                        else:
                            print(f' {hats[h]} ', end='')
                    print(guesses, end='')
                    match correct_bits:
                        case 1:
                            print('Only player 0', hats[3], hats[1])
                        case 2:
                            print('Only player 1', hats[0], hats[2])
                        case 4:
                            print('Only player 2', hats[1], hats[3])
                        case 8:
                            print('Only player 3', hats[2], hats[0])
                        case _:
                            print('Success')
        elif verbose:
            print('***', hats, guesses, "Failure")
    if verbose: print("Total successes:", successes)
    return successes

#########
# This version of check_all_hats() wil return 0 as soon as it finds a
# failure. It returns 81 if everything works.

def check_perfection(s0, s1, s2, s3, verbose=False):
    """s0~3 = 3x3 strategy for that player"""
    for hats in itertools.product(range(3), repeat=4):
        guesses = [s0[hats[3]][hats[1]],
                   s1[hats[0]][hats[2]],
                   s2[hats[1]][hats[3]],
                   s3[hats[2]][hats[0]]]
        if (hats[0] != guesses[0] and hats[1] != guesses[1] and
            hats[2] != guesses[2] and hats[3] != guesses[3]):
                return 0
    return 81

#########
# This function checks every way to distribute two strategies
# between the players. It found no successful strategies.

def iterate2():
    fifth = -1
    for raw_strat in itertools.product(range(3), repeat=18):
        if fifth != raw_strat[4]:
            print(f"Checking {raw_strat[0:5]}...")
            fifth = raw_strat[4]
        s0 = list(chunks(raw_strat[0:9], 3))
        s1 = list(chunks(raw_strat[9:18], 3))
        successes = check_perfection(s0, s1, s1, s1)
        if successes == 81:
            print('0', s0, '1, 2, 3', s1, successes)
        successes = check_perfection(s0, s0, s1, s1)
        if successes == 81:
            print('0, 1', s0, '2, 3', s1, successes)
        successes = check_perfection(s0, s1, s0, s1)
        if successes == 81:
            print('0, 2', s0, '1, 3', s1, successes)
            
# The best result of the above using the check_all_hats() version:
# 0 [(0, 0, 0), (1, 1, 2), (1, 0, 0)]
# 1, 2, 3 [(1, 0, 0), (1, 2, 2), (1, 0, 0)] 79

#########
# After running this long enough to see it wouldn't work,
# I switched to C++ for iterate3(). 

# This is the answer I got from my C++ code:

# check_all_hats([[0,0,1],[0,1,2],[1,2,2]],
#                [[2,2,1],[2,1,0],[1,0,0]],
#                [[1,0,0],[2,1,0],[2,2,1]],
#                [[1,0,0],[2,1,0],[2,2,1]], verbose=True)

#########
# I wrote this version of the test function as a final check
# of my answer. Instead of a 4x3x3 nested list of guesses, 
# I used the inputs each player would see and match/case'd
# their guess from that.

def test_english(verbose=False):
    """Strategies based on what each player sees."""
    successes = 0
    for hats in itertools.product(range(3), repeat=4):
        guesses = [-1, -1, -1, -1]
        # Simulating player 0
        match hats[3] + hats[1]:
            case 0 | 1:
                guesses[0] = 0
            case 2:
                guesses[0] = 1
            case 3 | 4:
                guesses[0] = 2
        # Simulating player 1
        match hats[0] + hats[2]:
            case 0 | 1:
                guesses[1] = 2
            case 2:
                guesses[1] = 1
            case 3 | 4:
                guesses[1] = 0
        # Simulating player 2
        match hats[1] - hats[3]:
            case -2 | -1:
                guesses[2] = 0
            case 0:
                guesses[2] = 1
            case 1 | 2:
                guesses[2] = 2
        # Simulating player 3
        match hats[2] - hats[0]:
            case -2 | -1:
                guesses[3] = 0
            case 0:
                guesses[3] = 1
            case 1 | 2:
                guesses[3] = 2
        if (hats[0] == guesses[0] or hats[1] == guesses[1] or
            hats[2] == guesses[2] or hats[3] == guesses[3]):
                successes += 1
                if verbose:
                    correct_bits = 0
                    for h in range(4):
                        if hats[h] == guesses[h]:
                            print(f'<{hats[h]}>', end='')
                            correct_bits += 2**h
                        else:
                            print(f' {hats[h]} ', end='')
                    print(guesses, end='')
                    match correct_bits:
                        case 1:
                            print('Only player 0', hats[3], hats[1])
                        case 2:
                            print('Only player 1', hats[0], hats[2])
                        case 4:
                            print('Only player 2', hats[1], hats[3])
                        case 8:
                            print('Only player 3', hats[2], hats[0])
                        case _:
                            print('Success')
        elif verbose:
            print('***', hats, guesses, "Failure")
    if verbose: print("Total successes:", successes)
    return successes
